### PR TITLE
Export ochttp.Handler to provide options on the type

### PR DESCRIPTION
**Pull request by rakyll**
Required for #426.

### CODE REVIEW

File: plugin/ochttp/httptrace.go [hunk 1]
```diff
@@ -132,34 +132,39 @@ func (t *traceTransport) CancelRequest(req *http.Request) {
 	}
 }
 
-// NewHandler returns a http.Handler from the given handler
-// that is aware of the incoming request's span.
-// The span can be extracted from the incoming request in handler
-// functions from incoming request's context:
+// NewHandler returns a new handler.
+func NewHandler(base http.HandlerFunc) *Handler {
```
Comment (odeke-em) [inline plugin/ochttp/httptrace.go]: Am afraid changing this signature from http.Handler to http.HandlerFunc will restrict the flexibility and interop with the outside world. It is easier to create an http.Handler than an http.HandlerFunc.
Let's make this
```go
func NewHandler(base http.Handler) *Handler
```
even the function suffix says "Handler" :)

File: plugin/ochttp/httptrace_test.go [hunk 2]
```diff
@@ -139,7 +139,7 @@ func TestHandler(t *testing.T) {
 		t.Run(tt.header, func(t *testing.T) {
 			propagator := &testPropagator{}
 
-			handler := NewHandler(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Comment (odeke-em) [inline plugin/ochttp/httptrace_test.go]: Please see my suggestion to keep `NewHandler(http.Handler)`

File: plugin/ochttp/httptrace.go (hunk 1 above)
Comment (ramonza) [reply plugin/ochttp/httptrace.go]: @odeke-em agreed if this is the recommended way to create a Handler then it's odd that it only takes a HandlerFunc not Handler.

File: plugin/ochttp/httptrace.go [hunk 3]
```diff
@@ -169,7 +174,7 @@ func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	span.SetAttributes(requestAttrs(r)...)
 
 	r = r.WithContext(ctx)
-	h.handler.ServeHTTP(w, r)
+	h.Handler.ServeHTTP(w, r)
```
Comment (ramonza) [inline plugin/ochttp/httptrace.go]: Should we make the zero value useful by having the default behavior be to call http.DefaultMux with defaultFormat propagation? If we do this, then I think we should also rename or remove NewHandler to address @odeke-em's concern.

**Comment by odeke-em** [comment]
@rakyll we finna need a rebase.

File: plugin/ochttp/httptrace.go (hunk 1 above)
Comment (rakyll) [reply plugin/ochttp/httptrace.go]: Done

File: plugin/ochttp/httptrace_test.go (hunk 2 above)
Comment (rakyll) [reply plugin/ochttp/httptrace_test.go]: Done

File: plugin/ochttp/trace.go [hunk 4]
```diff
@@ -132,34 +132,39 @@ func (t *traceTransport) CancelRequest(req *http.Request) {
...
 
-type handler struct {
-	handler http.Handler
-	format  propagation.HTTPFormat
+	// Handler is the handler used to handle the incoming request.
+	Handler http.Handler
 }
 
-func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+// TODO(jbd): Add Handler.NoTrace and Handler.NoStats.
```
Comment (odeke-em) [inline plugin/ochttp/trace.go]: Aha cool, I was actually just looking for the link of the code that @ramonza added to the client with NoTrace and NoStats.

File: plugin/ochttp/trace.go [hunk 5]
```diff
@@ -132,34 +132,39 @@ func (t *traceTransport) CancelRequest(req *http.Request) {
...
 // Incoming propagation mechanism is determined by the given HTTP propagators.
-func NewHandler(base http.Handler, format propagation.HTTPFormat) http.Handler {
-	if format == nil {
-		format = defaultFormat
-	}
-	return &handler{handler: base, format: format}
-}
+type Handler struct {
+	// Propagation defines how traces are propagated. If unspecified,
+	// B3 propagation will be used.
```
Comment (odeke-em) [inline plugin/ochttp/trace.go]: Where in the code do we fallback to B3 propagation if the zero value of Propagation is present?

File: plugin/ochttp/trace.go (hunk 4 above)
Comment (rakyll) [reply plugin/ochttp/trace.go]: Ack

File: plugin/ochttp/trace.go (hunk 5 above)
Comment (rakyll) [reply plugin/ochttp/trace.go]: Done

**Comment by rakyll** [comment]
PTAL!

File: plugin/ochttp/example_test.go [hunk 6]
```diff
@@ -26,3 +27,17 @@ func ExampleTransport() {
 	}
 	_ = client // use client to perform requests
 }
+
+var usersHandler http.Handler
+
+func ExampleHandler() {
+	// Enables OpenCensus for the default serve mux.
+	http.Handle("/users", usersHandler)
+	log.Fatal(http.ListenAndServe("localhost:8080", &ochttp.Handler{}))
```
Comment (odeke-em) [inline plugin/ochttp/example_test.go]: Nice!!

File: plugin/ochttp/trace.go (hunk 4 above)
Comment (odeke-em) [reply plugin/ochttp/trace.go]: Awesome, thank you!

File: plugin/ochttp/example_test.go [hunk 7]
```diff
@@ -26,3 +27,17 @@ func ExampleTransport() {
...
+}
+
+func ExampleHandler_mux() {
+	mux := http.NewServeMux()
+	mux.Handle("/users", usersHandler)
+	log.Fatal(http.ListenAndServe("localhost:8080", &ochttp.Handler{Handler: mux}))
```
Comment (odeke-em) [inline plugin/ochttp/example_test.go]: Could you add a Propagation here to exhibit that modification?

File: plugin/ochttp/example_test.go (hunk 6 above)
Comment (odeke-em) [inline plugin/ochttp/example_test.go]: This one looks alright without a Propagation, but let's modify the comment 
```go
// Enables OpenCensus for the default serve mux.
```
to say
```go
// Enables OpenCensus for the default serve mux and B3 Propagation, the defaults
```

File: plugin/ochttp/example_test.go (hunk 7 above)
Comment (rakyll) [reply plugin/ochttp/example_test.go]: Done

File: plugin/ochttp/example_test.go (hunk 6 above)
Comment (rakyll) [reply plugin/ochttp/example_test.go]: Done

File: plugin/ochttp/trace.go [hunk 8]
```diff
@@ -132,34 +132,43 @@ func (t *traceTransport) CancelRequest(req *http.Request) {
 	}
 }
 
-// NewHandler returns a http.Handler from the given handler
-// that is aware of the incoming request's span.
-// The span can be extracted from the incoming request in handler
-// functions from incoming request's context:
+// NewHandler returns a new handler.
+func NewHandler(base http.Handler) *Handler {
```
Comment (ramonza) [inline plugin/ochttp/trace.go]: Is this needed?

Comment (rakyll) [reply plugin/ochttp/trace.go]: Removed

**Pull request merged**